### DISCUSSION TITLE
[Feature Request]: Support using __call__ in `add_api_route` `endpoint` parameter

### DISCUSSION THREAD

**Discussion by sidekick-eimantas** [fastapi/fastapi#10238] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import pathlib

from fastapi import FastAPI
import pydantic
import uvicorn


class Hello:
    class Request(pydantic.BaseModel):
        id: str

    class Response(pydantic.BaseModel):
        hello: str

    def __init__(self, name: str) -> None:
        self.name = name

    async def __call__(self, request: Request) -> Response:
        return self.Response(hello=f"{request.id} {self.name}")


app = FastAPI()
handler = Hello(name="test")
app.add_api_route("/hello", handler, methods=["POST"])


def run() -> None:
    server_config = uvicorn.Config(
        app=app,
        host="localhost",
        port=8080,
        reload=True,
        reload_dirs=[str(pathlib.Path.cwd())],
        reload_includes=["*"],
        server_header=False,
        proxy_headers=True,
    )
    server = uvicorn.Server(config=server_config)
    server.run()


run()
```

### Description

The code above does not work, it raises `fastapi.exceptions.ResponseValidationError: 1 validation errors:
  {'type': 'model_attributes_type', 'loc': ('response',), 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': <coroutine object Hello.__call__ at 0x1051d1930>, 'url': 'https://errors.pydantic.dev/2.3/v/model_attributes_type'}`. Seems that fastapi did not recognize the class as an async callable

A simple workaround is to pass in the `handler.__call__` reference explicitly.

This might be related: https://github.com/encode/starlette/issues/886

This isn't a huge deal and probably shouldn't spend time on it, given the workaround is very simple, unless the implementation is trivial

### Operating System

macOS

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.7

**Comment by danielfcollier** [comment]
I could reproduce your problem. After some investigation (and learning), I've found this is not a FastAPI problem, it is a problem with the definition of the caller method, it shouldn't be `async`, because it is not an asynchronous call (at least this example implementation).

Please, try again changing to:

```python
class Hello:
    class Request(pydantic.BaseModel):
        id: str

    class Response(pydantic.BaseModel):
        hello: str

    def __init__(self, name: str) -> None:
        self.name = name

    def __call__(self, request: Request) -> Response:
        return self.Response(hello=f"{request.id} {self.name}")
```

**Comment by danielfcollier** [comment]
I've realized that you might really need an asynchronous call, I couldn't find a straightforward solution, but this might solve your problem:

```python
import asyncio
from fastapi import FastAPI
import pydantic

class Hello:
    class Request(pydantic.BaseModel):
        id: str

    class Response(pydantic.BaseModel):
        hello: str

    def __init__(self, name: str) -> None:
        self.name = name

    async def __call__(self, request: Request) -> Response:
        await asyncio.sleep(1)
        return self.Response(hello=f"{request.id} {self.name}")


async def handler(request: Hello.Request) -> Hello.Response:
  _handler = Hello(name="test")
  return await _handler(request)


app = FastAPI()
app.add_api_route("/hello", handler, methods=["POST"])
```

**Comment by tiangolo** [comment, marked as answer]
Thanks for the interest! So, the way you are declaring your Pydantic models as class attributes inside of a class is not really supported in Pydantic nor FastAPI. It would probably be better to refactor that.

Is there any reason why you need the code with that structure?

**Reply by alukach** [reply]
I think using a class can make sense for code organization, typically with passing configuration (determined when the app loads) to handlers.  For example:

```py
# app.py
from pydantic_settings import BaseSettings
from fastapi import FastAPI
from .handlers import AddVal

class Settings(BaseSettings):
    val: int

settings = Settings()  # initialized at boot time
app = FastAPI()

app.add_api_route(
  '/add',
  AddVal(val=settings.val),
  methods=["GET"],
)
```

```py
# handlers.py
from dataclasses import dataclass

@dataclass
class AddVal:
  val: int

  def __call__(self, val: int):
    return self.val + val
```

This way, I can configure my routes in the `app.py` but define them elsewhere.  This is helpful for avoiding circular dependencies (obviously `handlers.py` couldn't import `app.py:settings`).  I usually try to keep one file where I initialize everything `app.py`, typically using a factory function; and so initializing my handlers within that file seems to make sense to me.

Some workarounds:

* use `app.state` to pass along configuration to handlers. This seems like it could be a bit of an anti-pattern in the FastAPI world as the docs suggest using dependencies instead.
* name method something other than `__call__`, like `dispatch`. Slightly less clean, but it works:
    ```py
    # app.py
    from pydantic_settings import BaseSettings
    from fastapi import FastAPI
    from .handlers import AddVal
    
    class Settings(BaseSettings):
        val: int
    
    settings = Settings()  # initialized at boot time
    app = FastAPI()
    
    app.add_api_route(
      '/add',
      AddVal(val=settings.val).dispatch,
      methods=["GET"],
    )
    ```
    
    ```py
    # handlers.py
    from dataclasses import dataclass
    
    @dataclass
    class AddVal:
      val: int
    
      def dispatch(self, val: int):
        return self.val + val
    ```
